securedrop/journalist_app/api.py

```diff
@@ -0,0 +1,250 @@
+import hashlib
+import json
+from werkzeug.exceptions import default_exceptions  # type: ignore
+
+from flask import abort, Blueprint, current_app, jsonify, request, send_file
+
+from db import db
+from journalist_app import utils
+from models import (Journalist, Reply, Source, Submission,
+                    LoginThrottledException, InvalidUsernameException,
+                    BadTokenException, WrongPasswordException)
+from store import NotEncrypted
+
+
+TOKEN_EXPIRATION_MINS = 60 * 8
+
+
+def get_user_object(request):
+    """Helper function to use in token_required views that need a user
+    object
+    """
+    auth_token = request.headers.get('Authorization').split(" ")[1]
+    user = Journalist.validate_api_token_and_get_user(auth_token)
+    return user
+
+
+def token_required(f):
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        try:
+            auth_header = request.headers['Authorization']
+        except KeyError:
+            return abort(403, 'API token not found in Authorization header.')
+
+        if auth_header:
+            auth_token = auth_header.split(" ")[1]
+        else:
+            auth_token = ''
+        if not Journalist.validate_api_token_and_get_user(auth_token):
+            return abort(403, 'API token is invalid or expired.')
+        return f(*args, **kwargs)
+    return decorated_function
+
+
+def get_or_404(model, object_id, column=''):
+    if column:
+        result = model.query.filter(column == object_id).one_or_none()
+    else:
+        result = model.query.get(object_id)
+    if result is None:
+        abort(404)
+    return result
+
+
+def make_blueprint(config):
+    api = Blueprint('api', __name__)
+
+    @api.route('/')
+    def get_endpoints():
+        endpoints = {'sources_url': '/api/v1/sources',
+                     'current_user_url': '/api/v1/user',
+                     'submissions_url': '/api/v1/submissions',
+                     'auth_token_url': '/api/v1/token'}
+        return jsonify(endpoints), 200
+
+    @api.route('/token', methods=['POST'])
+    def get_token():
+        creds = json.loads(request.data)
+
+        username = creds.get('username', None)
+        passphrase = creds.get('passphrase', None)
+        one_time_code = creds.get('one_time_code', None)
+
+        if username is None:
+            return abort(400, 'username field is missing')
+        if passphrase is None:
+            return abort(400, 'passphrase field is missing')
+        if one_time_code is None:
+            return abort(400, 'one_time_code field is missing')
+
+        try:
+            journalist = Journalist.login(username, passphrase, one_time_code)
+            token_expiry = datetime.utcnow() + timedelta(
+                seconds=TOKEN_EXPIRATION_MINS * 60)
+            response = jsonify({'token': journalist.generate_api_token(
+                 expiration=TOKEN_EXPIRATION_MINS * 60),
+                 'expiration': token_expiry.isoformat() + 'Z'})
```

Comment:
Nit: This is not incorrect, but if the timezone is ever not UTC, this will return an invalid string. It would be marginally safer to use `strftime`.

Reply:
well, since this is constructed from a timezone that is always UTC, since it's set as `datetime.utcnow()` above, this should be safe as is no?

I think the only time we need to handle non-UTC timezones is in presenting timezones to the journalist user, i.e. when we expose datetimes in the UI, otherwise... we are all UTC all the time (see what I did there)

Reply:
Fair enough. This is a nit, and def not a blocker.